File: pr1047e/remove_duplicates.cpp
// g++ -std=c++11 *.cpp -o test && ./test && rm -f test
#include <cassert>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    string removeDuplicates(string s)
    {
        string ans = "";
        for (char c : s) {
            if (ans.size() == 0 || ans.back() != c) {
                ans.push_back(c);
            } else {
                ans.pop_back();
            }
        }
        return ans;
    }
};

// TESTS
int main(int argc, char* argv[])
{
    vector<vector<string>> tests = {
        { "abcd", "abcd" },
        { "abbaca", "ca" }
    };
    Solution sol;
    for (auto& t : tests) {
        string actual = sol.removeDuplicates(t[0]);
        cout << "Remove duplicates in " << t[0] << " -> " << actual << endl;
        assert(actual == t[1]);
    }
    return 0;
}

File: pr1047e/remove_duplicates.py
class Solution:
    def removeDuplicates(self, s: str) -> str:
        ans = []
        for c in s:
            if ans and ans[-1] == c:
                ans.pop()
            else:
                ans.append(c)
        return "".join(ans)


# TESTS
for s, expected in [
    ("abcd", "abcd"),
    ("abbaca", "ca"),
    ("azxxzy", "ay"),
]:
    sol = Solution()
    actual = sol.removeDuplicates(s)
    print("Remove duplicates in", s, "->", actual)
    assert actual == expected
